plugin/output/s3/structs.go

```diff
@@ -0,0 +1,144 @@
+	BucketNameFieldForTrack string `json:"bucket_name_field_for_track"` //*
+
+	//> @3@4@5@6
+	//> Name of s3url in commiter message.
+	S3UrlNameFieldForTrack string `json:"s3_url_name_field_for_track"` //*
+
+	//> @3@4@5@6
+	//> Constant part of commit, must be valid json fields.
+	//> Useful to add some contant data to commiter message: `"service": "serviceName", "release": "v1.0.1"`
+	ConstantCommitMessagePart string `json:"contant_message_part"` //*
+	//> @3@4@5@6
+	KafkaCfg kafka.Config `json:"kafka_config" child:"true"` //*
+}
+
+type Plugin struct {
+	controller    pipeline.OutputPluginController
+	logger        *zap.SugaredLogger
+	config        *Config
+	params        *pipeline.OutputPluginParams
+	fileExtension string
+
+	// defaultClient separated from others to avoid locks with dynamic buckets.
+	defaultClient ObjectStoreClient
+	clients       map[string]ObjectStoreClient
+	limiter       *ObjectStoreClientLimiter
+
+	outPlugins            *file.Plugins
+	dynamicPlugCreationMu sync.Mutex
+
+	compressCh chan fileDTO
+	uploadCh   chan fileDTO
+	compressor compressor
+
+	commitMode      bool
+	commiterWrapper *CommiterWrapper
+	inputController pipeline.InputPluginController
+}
+
+//! config-params
+//^ config-params
+type Config struct {
+	//> @3@4@5@6
+	//> Under the hood this plugin uses /plugin/output/file/ to collect logs
+	FileConfig file.Config `json:"file_config" child:"true"` //*
+
+	//> @3@4@5@6
+	//> Compression type
+	CompressionType string `json:"compression_type" default:"zip" options:"zip"` //*
+
+	// s3 section
+
+	//> @3@4@5@6
+	//> Endpoint address of default bucket.
+	Endpoint string `json:"endpoint" required:"true"` //*
+	//> @3@4@5@6
+	//> s3 access key.
+	AccessKey string `json:"access_key" required:"true"` //*
+	//> @3@4@5@6
+	//> s3 secret key.
+	SecretKey string `json:"secret_key" required:"true"` //*
+	//> @3@4@5@6
+	//>  s3 default bucket.
+	DefaultBucket string `json:"bucket" required:"true"` //*
+	//> @3@4@5@6
+	//> MultiBuckets is additional buckets, which can also receive event.
+	//> Event must contain `bucket_name` field which value is s3 bucket name.
+	//> Events without `bucket_name` sends to DefaultBucket.
+	MultiBuckets `json:"multi_buckets" required:"false"` //*
+	//> @3@4@5@6
+	//> s3 connection secure option.
+	Secure bool `json:"secure" default:"false"` //*
+	//> @3@4@5@6
+	//> BucketEventField field change destination bucket of event to fields value.
+	//> Fallback to DefaultBucket if BucketEventField bucket doesn't exist.
+	BucketEventField string `json:"bucket_field_event" default:""` //*
+	//> @3@4@5@6
+	//> DynamicBucketsLimit regulates how many buckets can be created dynamically.
+	//> Prevents problems when some random strings in BucketEventField where
+	DynamicBucketsLimit int `json:"dynamic_buckets_limit" default:"32"` //*
+
+	// 2 phase commit section
+	//> @3@4@5@6
+	//> Describes config of commit.
+	CommitCfg *commitConfig `json:"commit_config" child:"true"` //*
```

Comment:
Why is there a pointer? ?

Reply:
It's optional. Otherwise cfg parser will fail on default kafka plugin - on doesn't have broker list etc